[data-client] Add structured logging to data client

khiemngo opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

This PR adds structured logging to state-sync's data client.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

- Re-ran all tests under `diem/state-sync/` and all tests passed. This is to ensure the added logging does not affect existing tests.
- Manually check some logging printed on the terminal.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

khiemngo commented:
Hi @JoshLind , could you please give me some initial feedback on the structured logging (e.g., log entry and event naming, what to log, etc.) when you have time? Thank you! 

PS: I did also take a look at state-sync-v1 and state-sync-v2 logging to learn if there is any naming convention to follow and what to include in a log entry. I was also thinking about logging more data (e.g., peer score after being updated, storage and data client configs) but I'm not quite sure if it is very helpful at the moment.

khiemngo commented on state-sync/diem-data-client/src/diemnet/logging.rs:
```diff
@@ -0,0 +1,47 @@
+    error: Option<&'a Error>,
+    event: Option<LogEvent>,
+    message: Option<&'a str>,
+}
```
Is there any existing data client ID field that we can use to identify a log entry?

gregnazario replied:
Is this to identify the machine or the individual client on the server(is there supposed to be more than one?)

JoshLind replied:
@khiemngo, there is no data client ID field in this context (I don't think we really need it). Especially as users of the diem data client will also do their own logging, and the data client relies on the network implementation to handle request/response mappings.

gregnazario commented on state-sync/diem-data-client/src/diemnet/logging.rs:
```diff
@@ -0,0 +1,47 @@
+    name: LogEntry,
+    error: Option<&'a Error>,
+    event: Option<LogEvent>,
+    message: Option<&'a str>,
```
I don't think you'll need a `message` field, since structured logging already provides this.

You can log e.g.

```
info!(Schema::new(...), "Here is my message with {} field", field);
```

This will handle formatting, and it'll be named as `message`.

JoshLind replied:
I think this is a matter of preference, but good to know that it'll automatically parse the string.

gregnazario commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -434,17 +441,38 @@ impl DataSummaryPoller {
             self.data_client.update_summary(peer, storage_summary);
             self.data_client.update_global_summary_cache();
+
+            info!(
+                (LogSchema::new(LogEntry::PeerStates)
+                    .event(LogEvent::AggregateSummary)
+                    .message(&format!(
+                        "Global data summary: {:?}",
+                        self.data_client.get_global_data_summary()
+                    )))
+            );
```
Unless this information would be used to determine an action, I'd keep anything that occurs on a regular basis to a `debug` level unless it's a simple message without too much variance in it.

JoshLind commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -434,17 +441,38 @@ impl DataSummaryPoller {
+                    .event(LogEvent::AggregateSummary)
+                    .message(&format!(
+                        "Global data summary: {:?}",
+                        self.data_client.get_global_data_summary()
```
Following from my comment above, I think it makes sense for us to log this (maybe `debug` level)? But I also have concerns about size. Specifically, if you look at the `GlobalDataSummary` object, it contains `synced_ledger_infos` in the `AdvertisedData`. When we `Debug` format those, they can be massive. To avoid this, we should override the `Debug` formatting for `GlobalDataSummary` ourselves so that we don't print those out entirely.  Feel free to do this in a follow up PR and we can sync :)

khiemngo replied:
I'm going to implement overriding `Debug` for `GlobalDataSummary` and other responses with potentially large payload for other storage request types in a follow-up PR.

JoshLind reviewed (commented):
Thanks @khiemngo! Great start 🚀 

There's a few other things I'd like us to log:
1. Each request/response along with the peer. We might be able to get away with logging the request data as well (if it's small), but we probably shouldn't log all the response data (because it might be huge).
2. Any errors that occur in the rest of the diem data client. These are important to know about.
3. When a peer is ignored (i.e., the score drops below the threshold) and when an ignored peer is no longer ignored (i.e., the score rises above the threshold).

I can provide pointers for these, but I'd like to leave it as an exercise for the reader (so you can get some familiarity with the code) 😄

JoshLind commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -434,17 +441,38 @@ impl DataSummaryPoller {
-                        "Error encountered when polling peer ({:?})! Error: {:?}",
-                        peer,
-                        error
+                    debug!(
```
@khiemngo, yeah, my thinking is that a bad peer can only do this so many times before they're ignored 😄

khiemngo replied:
Thanks for the pointer to `sample!` @gregnazario! (something new I learnt today!). Will keep it in mind if logging here causes a problem in future.

khiemngo commented on state-sync/diem-data-client/src/diemnet/state.rs:
```diff
@@ -118,14 +122,34 @@ impl PeerStates {
+        let old_score = self.inner.entry(peer).or_default().score;
+        self.inner.entry(peer).or_default().update_score_success();
+        let new_score = self.inner.entry(peer).or_default().score;
+        if old_score <= IGNORE_PEER_THRESHOLD && new_score > IGNORE_PEER_THRESHOLD {
```
I could have gone with just logging the score whenever it is updated or as long as it's above/below the threshold. However, it seems that we are more interested in the events when a peer switches between ignored/no longer ignored. Please let me know if we just want to log the score and I can change.

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -434,17 +482,28 @@ impl DataSummaryPoller {
-                        "Error encountered when polling peer ({:?})! Error: {:?}",
-                        peer,
-                        error
+                    error!(
```
I also log the errors inside `send_request_to_peer()` when a response callback encounters an error, so this logging here may be not necessary anymore.

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -247,6 +281,15 @@ impl DiemNetDataClient {
+                        .event(LogEvent::ResponseError)
+                        .request_type(request.get_label())
+                        .request_id(id)
+                        .peer(&peer)
```
perhaps add some `.message(..)` here?

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -210,6 +235,15 @@ impl DiemNetDataClient {
 
         match result {
             Ok(response) => {
+                // TODO(khiemngo): consider logging response payload
```
will probably do this in a follow-up PR.

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -415,13 +417,18 @@ impl DataSummaryPoller {
-                    trace!("Unable to select the next peer! Error: {:?}", error);
+                    debug!(
+                        (LogSchema::new(LogEntry::StorageSummaryRequest)
+                            .event(LogEvent::MissingPeers)
```
Changed `MissingPeers` to `NoPeersToPoll`

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -415,13 +417,18 @@ impl DataSummaryPoller {
                 Ok(peer) => peer,
                 Err(error) => {
-                    trace!("Unable to select the next peer! Error: {:?}", error);
+                    debug!(
```
Used `.error(...)` style.

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -434,17 +441,38 @@ impl DataSummaryPoller {
-                        "Error encountered when polling peer ({:?})! Error: {:?}",
-                        peer,
-                        error
+                    debug!(
```
Used `.error(...)` style.

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -396,8 +400,6 @@ impl DataSummaryPoller {
     }
 
     pub async fn start(self) {
-        trace!("Starting the diem data poller!");
```
Log this at `info` level log under the entry `LogEntry::DataSummaryPollerStart`

khiemngo commented:
> Thanks @khiemngo! Great start 🚀
> 
> There's a few other things I'd like us to log:
> 
> 1. Each request/response type along with the peer (at `debug` level). We might be able to get away with logging the request data as well (if it's small), but we probably shouldn't log all the response data (because it might be huge).
> 2. Any errors that occur in the rest of the diem data client. These are important to know about.
> 3. When a peer is ignored (i.e., the score drops below the threshold) and when an ignored peer is no longer ignored (i.e., the score rises above the threshold).
> 
> I can provide pointers for these, but I'd like to leave it as an exercise for the reader (so you can get some familiarity with the code) 😄 Let me know if you get blocked!

Thanks for the feedback @JoshLind . I've tried to add more logging based on your suggestions (1-3). I've decided to implement overriding `Debug` for `GlobalDataSummary` (and perhaps also for some other response types) in the next PR. Please let me know if I should make further changes. Thank you!

JoshLind commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -434,17 +482,28 @@ impl DataSummaryPoller {
+                    error!(
+                        (LogSchema::new(LogEntry::StorageSummaryResponse)
+                            .event(LogEvent::PeerPollingError)
+                            .message(&format!("Error encountered when polling peer {:?}", peer))
```
We should probably log the `peer` as the same way as above, right? e.g., 
```
...
.message(....)
.peer(peer)
...
```

khiemngo replied:
Used `.peer` field.

JoshLind commented on state-sync/diem-data-client/src/diemnet/state.rs:
```diff
@@ -118,14 +122,34 @@ impl PeerStates {
+            debug!(
+                (LogSchema::new(LogEntry::PeerStates)
+                    .event(LogEvent::PeerNoLongerIgnored)
+                    .message(&format!("Peer {:?} will no longer be ignored", peer))
```
Likewise, can we not use the `.peer(...)` entry to log the peer.

khiemngo replied:
Used .peer field.

JoshLind reviewed (commented):
Thanks @khiemngo! Various comments (mostly nits), but it looks great! Once you address them, this should be good to go! 🚀

khiemngo commented on state-sync/diem-data-client/src/diemnet/logging.rs:
```diff
@@ -0,0 +1,75 @@
+#[derive(Clone, Copy, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum LogEntry {
+    StorageServiceRequest,
```
Sorted them in alphabetical order.

khiemngo commented on state-sync/diem-data-client/src/diemnet/logging.rs:
```diff
@@ -0,0 +1,75 @@
+#[derive(Clone, Copy, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum LogEvent {
+    // StorageServiceRequest events
```
Sorted them in alphabetical order.

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -163,7 +167,18 @@ impl DiemNetDataClient {
         E: Into<Error>,
     {
-        let peer = self.choose_peer(&request)?;
+        let peer = match self.choose_peer(&request) {
```
Used `map_err` style.

Thanks! It works! I'll try to get used to using `map_` from now on.

khiemngo commented on state-sync/diem-data-client/src/diemnet/logging.rs:
```diff
@@ -0,0 +1,75 @@
+pub enum LogEvent {
+    // StorageServiceRequest events
+    SendRequest,
+    PeerSelectingError,
```
Changed to `PeerSelectionError`

khiemngo commented on state-sync/diem-data-client/src/diemnet/mod.rs:
```diff
@@ -199,6 +214,16 @@ impl DiemNetDataClient {
     ) -> Result<Response<StorageServiceResponse>, Error> {
         let id = self.next_response_id();
+
+        info!(
```
Used `debug!`

khiemngo commented on state-sync/diem-data-client/src/diemnet/state.rs:
```diff
@@ -118,14 +122,34 @@ impl PeerStates {
+                (LogSchema::new(LogEntry::PeerStates)
+                    .event(LogEvent::PeerNoLongerIgnored)
+                    .message(&format!("Peer {:?} will no longer be ignored", peer))
+                    .peer_score(new_score))
```
Removed logging for peer score.

@JoshLind Yeah I don't think it adds much value to the logging. 
(I previously thought whoever wrote/designed the scoring schemes might be interested in seeing how far from the threshold the new score is after being updated...)

khiemngo commented:
Thanks @JoshLind for additional feedback. I addressed all nits. If there are more nits to address, please let me know. Thank you!!

khiemngo commented:
> Thanks @khiemngo! Various comments (mostly nits), but it looks great! Once you address them, this should be good to go! 🚀

Thanks @JoshLind for additional feedback. I addressed all nits. If there are more nits to address, please let me know. Thank you!!

JoshLind reviewed (approved):
Awesome, thanks @khiemngo! ❤️

khiemngo commented:
/land
